File: tests/test_cross.py
from fractions import Fraction

from rat_trig.trigonom import cross


def test_cross() -> None:
    """Test cross product"""
    # Test with integer vectors
    v_1_int1 = [1, 2]
    v_2_int1 = [3, 4]
    assert cross(v_1_int1, v_2_int1) == -2

    # Test with float vectors
    v_1_float = [1.0, 2.0]
    v_2_float = [3.0, 4.0]
    assert cross(v_1_float, v_2_float) == -2.0

    # Test with Fraction vectors
    v_1_frac = [Fraction(1, 2), Fraction(1, 4)]
    v_2_frac = [Fraction(1, 6), Fraction(1, 8)]
    assert cross(v_1_frac, v_2_frac) == Fraction(1, 48)

    # Test with integer vectors (parallel)
    v_1_int2 = [1, 2]
    v_2_int2 = [1, 2]
    assert cross(v_1_int2, v_2_int2) == 0

    # Test with integer vectors (perpendicular)
    v_1_int3 = [1, 0]
    v_2_int3 = [0, 1]
    assert cross(v_1_int3, v_2_int3) == 1


if __name__ == "__main__":
    import pytest

    pytest.main()

File: cpp_ai/tests/test_cross.cpp
/**
 * @file test_cross.cpp
 * @brief Tests for cross product function using doctest
 */

#include "doctest.h"
#include "rat_trig/trigonom.hpp"

TEST_CASE("Testing cross function") {
    SUBCASE("Test with integer vectors") {
        std::array<int, 2> vec1 = {1, 2};
        std::array<int, 2> vec2 = {3, 4};
        CHECK(rat_trig::cross(vec1, vec2) == -2);
    }

    SUBCASE("Test with float vectors") {
        std::array<float, 2> vec1 = {1.0f, 2.0f};
        std::array<float, 2> vec2 = {3.0f, 4.0f};
        CHECK(rat_trig::cross(vec1, vec2) == doctest::Approx(-2.0f));
    }

    SUBCASE("Test with double vectors") {
        std::array<double, 2> vec1 = {1.0, 2.0};
        std::array<double, 2> vec2 = {3.0, 4.0};
        CHECK(rat_trig::cross(vec1, vec2) == doctest::Approx(-2.0));
    }

    SUBCASE("Test with integer vectors (parallel)") {
        std::array<int, 2> vec1 = {1, 2};
        std::array<int, 2> vec2 = {1, 2};
        CHECK(rat_trig::cross(vec1, vec2) == 0);
    }

    SUBCASE("Test with integer vectors (perpendicular)") {
        std::array<int, 2> vec1 = {1, 0};
        std::array<int, 2> vec2 = {0, 1};
        CHECK(rat_trig::cross(vec1, vec2) == 1);
    }

    SUBCASE("Test with fractions (using doubles)") {
        std::array<double, 2> vec1 = {1.0 / 2.0, 1.0 / 4.0};
        std::array<double, 2> vec2 = {1.0 / 6.0, 1.0 / 8.0};
        double expected = 1.0 / 48.0;
        CHECK(rat_trig::cross(vec1, vec2) == doctest::Approx(expected).epsilon(1e-10));
    }
}
